x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/HoistRemoteEnrichTopN.java

```diff
@@ -0,0 +1,168 @@
+import org.elasticsearch.xpack.esql.plan.logical.TopN;
+import org.elasticsearch.xpack.esql.plan.logical.UnaryPlan;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * Locate any TopN that is "visible" under remote ENRICH, and make a copy of it above the ENRICH,
+ * while making a copy of the original fields. Mark the original TopN as local.
+ * This is the same idea as {@link HoistRemoteEnrichLimit} but for TopN instead of Limit.
+ * This must happen after {@link ReplaceLimitAndSortAsTopN}.
+ * This also has handling for the case where Enrich overrides the field(s) TopN is using. In this case,
+ * we need to create aliases for the fields used by TopN, and then use those aliases in the copy TopN.
+ * Then we need to add Project to remove the alias fields. Fortunately, PushDownUtils has the logic to do that.
+ */
+public final class HoistRemoteEnrichTopN extends OptimizerRules.OptimizerRule<Enrich> implements OptimizerRules.CoordinatorOnly {
+    public HoistRemoteEnrichTopN() {
+        super(OptimizerRules.TransformDirection.UP);
+    }
+
+    @Override
+    protected LogicalPlan rule(Enrich en) {
+        if (en.mode() == Enrich.Mode.REMOTE) {
+            LogicalPlan plan = en.child();
+            List<Attribute> outputs = en.output();
+            // This loop only takes care of one TopN. Repeated TopNs may be a problem, we can't handle them here.
+            while (true) {
+                if (plan instanceof TopN topN && topN.local() == false) {
+                    Set<Attribute> missingAttributes = checkOrderInOutputs(topN.order(), outputs);
+
+                    // Do we need to do any aliasing because some of the fields were overridden by Enrich or other generating plans?
+                    if (missingAttributes.isEmpty()) {
+                        // No need for aliasing - then it's simple, just copy the TopN on top of Enrich and mark the original as local
+                        LogicalPlan transformedEnrich = en.transformDown(TopN.class, t -> t == topN ? topN.withLocal(true) : t);
+                        return new TopN(topN.source(), transformedEnrich, topN.order(), topN.limit(), false);
+                    } else {
+                        Set<String> missingNames = new HashSet<>(Expressions.names(missingAttributes));
+                        AttributeMap.Builder<Alias> aliasesForReplacedAttributesBuilder = AttributeMap.builder();
+                        List<Expression> rewrittenExpressions = new ArrayList<>();
+
+                        // Replace the missing attributes in order expressions with their aliases
+                        for (Order expr : topN.order()) {
+                            rewrittenExpressions.add(expr.transformUp(Attribute.class, attr -> {
+                                if (missingNames.contains(attr.name())) {
```

Comment:
nit: comparing by name isn't wrong here, but usually we use an `AttributeSet`, where `.contains` works by NameId. That'd avoid the dance where we start with a `Set<Attribute>` (not the same!) and then create a `Set<String>` from it.